pkg/csr/csr.go

```diff
@@ -0,0 +1,236 @@
+	log               logr.Logger
+}
+
+// NewCSRApprover returns a pointer to the CSRApprover
+func NewCSRApprover(k8sclientSet *kubernetes.Clientset, instanceAddresses []string, log logr.Logger) (*CSRApprover, error) {
+	if len(instanceAddresses) == 0 || k8sclientSet == nil {
+		return nil, errors.New("instance address list or kubernetes clientset should not be nil")
+	}
+	return &CSRApprover{k8sclientSet, instanceAddresses, log}, nil
+}
+
+// validateCSRContents returns true if the CSR request contents are valid
+func (c *CSRApprover) validateCSRContents(csr *certificatesv1.CertificateSigningRequest) (bool, error) {
+	parsedCSR, err := ParseCSR(csr.Spec.Request)
+	if err != nil {
+		return false, errors.Wrapf(err, "error parsing CSR %s", csr.Name)
+	}
+
+	dnsAddr := strings.TrimPrefix(parsedCSR.Subject.CommonName, NodeUserPrefix)
+	if dnsAddr == "" {
+		return false, errors.Errorf("CSR subject name does not contain the required node user prefix %s", csr.Name)
+	}
+
+	// check DNS of node requesting CSR matches instance IP/DNS address
+	var hasMatch bool
+	for _, instanceAdd := range c.instanceAddresses {
+		if areAddressesEqual(instanceAdd, dnsAddr) {
+			hasMatch = true
+			break
+		}
+	}
+	if !hasMatch {
+		// CSR is not from a BYOH Windows instance, don't deny it or return error
+		// since it might be from a linux node.
+		return false, nil
+	}
+	// Kubelet on a node needs two certificates for its normal operation:
+	// Client certificate for securely communicating with the Kubernetes API server
+	// Server certificate for use by Kubernetes API server to talk back to kubelet
+	// Both types are validated based on their contents
+	var validated bool
+	if isNodeClientCert(csr, parsedCSR) {
+		// Node client bootstrapper CSR is received before the instance becomes a node
+		// hence we should not proceed if a corresponding node already exists
+		_, err := c.k8sClientSet.CoreV1().Nodes().Get(context.TODO(), dnsAddr, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return false, errors.Wrapf(err, "unable to get node %s", dnsAddr)
+		} else if err == nil {
+			return false, errors.Wrapf(err, "%s node already exists, cannot validate CSR", dnsAddr)
+		}
+	} else {
+		if validated, err = c.validateKubeletServingCSR(csr, parsedCSR); !validated {
+			return false, errors.Wrapf(err, "unable to validate kubelet serving CSR %s", csr.Name)
+		}
+	}
+	return true, nil
+}
+
+// validateKubeletServingCSR validates a kubelet serving CSR for its contents
+func (c *CSRApprover) validateKubeletServingCSR(req *certificatesv1.CertificateSigningRequest, csr *x509.CertificateRequest) (bool, error) {
+	if req == nil || csr == nil {
+		return false, errors.New("CSR or request should not be nil")
+	}
+
+	// Check groups, we need at least: system:nodes, system:authenticated
+	if len(req.Spec.Groups) < 2 {
+		return false, errors.Errorf("CSR contains too few groups: %s", req.Name)
+	}
+	groupSet := sets.NewString(req.Spec.Groups...)
```

Comment:
I suggest naming this `groups`. It is not good practice to reflect the data structure being used in the variable name.